File: src/main/java/com/ai2dev/demo_travelcard_sb/dto/CreateTravelcardRequest.java
package com.ai2dev.demo_travelcard_sb.dto;

import com.ai2dev.demo_travelcard_sb.model.CardholderType;
import com.ai2dev.demo_travelcard_sb.model.TravelcardType;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import java.time.OffsetDateTime;
import java.util.List;

public record CreateTravelcardRequest(
    @NotNull
    TravelcardType travelcardType,

    @NotNull
    OffsetDateTime travelcardValidFrom,

    @NotNull
    OffsetDateTime travelcardValidTo,

    @Size(max = 255)
    @Pattern(regexp = "^[A-Za-z0-9 ]*$")
    String travelcardName,

    @NotBlank
    @Size(min = 11, max = 22)
    @Pattern(regexp = "^[A-Za-z0-9]+$")
    String travelcardNumber,

    @NotNull
    OffsetDateTime travelcardRequestedDate,

    @NotBlank
    @Size(min = 15, max = 15)
    String travelcardTransactionReference,

    OffsetDateTime travelcardUsableTo,

    @NotEmpty
    @Valid
    List<CardholderDto> cardholders
) {}

File: Models/CreateTravelCardRequest.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace TravelCardFunctionApp.Models;

public class CreateTravelCardRequest
{
    [JsonPropertyName("travelcardType")]
    public TravelcardType TravelcardType { get; set; }

    [JsonPropertyName("travelcardValidFrom")]
    public DateTime TravelcardValidFrom { get; set; }

    [JsonPropertyName("travelcardValidTo")]
    public DateTime TravelcardValidTo { get; set; }

    [JsonPropertyName("travelcardName")]
    public string? TravelcardName { get; set; }

    [JsonPropertyName("travelcardNumber")]
    public string TravelcardNumber { get; set; } = string.Empty;

    [JsonPropertyName("travelcardRequestedDate")]
    public DateTime TravelcardRequestedDate { get; set; }

    [JsonPropertyName("travelcardTransactionReference")]
    public string TravelcardTransactionReference { get; set; } = string.Empty;

    [JsonPropertyName("travelcardUsableTo")]
    public DateTime? TravelcardUsableTo { get; set; }

    [JsonPropertyName("cardholders")]
    public List<CardholderRequest> Cardholders { get; set; } = new();
}
